cmd/mlab_service_discovery/main.go

```diff
@@ -0,0 +1,224 @@
+	d.client, err = google.DefaultClient(oauth2.NoContext, scopes...)
+	if err != nil {
+		return nil, fmt.Errorf("Error setting up AppEngine client: %s", err)
+	}
+
+	// Create a new AppEngine service instance.
+	d.apis, err = appengine.New(d.client)
+	if err != nil {
+		return nil, fmt.Errorf("Error setting up AppEngine client: %s", err)
+	}
+
+	return d, nil
+}
+
+// getLabels creates a target configuration for a prometheus service discovery
+// file. The given service version should have a "SERVING" status, the instance
+// should be in a "RUNNING" state and have at least one forwarded port.
+//
+// In serialized form, the label set look like:
+//   {
+//       "labels": {
+//           "__aef_instance": "aef-etl--parser-20170418t195100-abcd",
+//           "__aef_max_total_instances": "20",
+//           "__aef_project": "mlab-sandbox",
+//           "__aef_public_protocol": "tcp",
+//           "__aef_service": "etl-parser",
+//           "__aef_version": "20170418t195100",
+//           "__aef_vm_debug_enabled": "true"
+//       },
+//       "targets": [
+//           "104.196.220.184:9090"
+//       ]
+//   }
+func getLabels(service *appengine.Service, version *appengine.Version, instance *appengine.Instance) map[string]interface{} {
+	labels := map[string]string{
+		aefLabelProject:      *project,
+		aefLabelService:      service.Id,
+		aefLabelVersion:      version.Id,
+		aefLabelInstance:     instance.Id,
+		aefMaxTotalInstances: fmt.Sprintf("%d", version.AutomaticScaling.MaxTotalInstances),
+		aefVmDebugEnabled:    fmt.Sprintf("%t", instance.VmDebugEnabled),
+	}
+	if strings.HasSuffix(version.Network.ForwardedPorts[0], "/udp") {
+		labels[aefLabelPublicProto] = "udp"
+	} else if strings.HasSuffix(version.Network.ForwardedPorts[0], "/tcp") {
+		labels[aefLabelPublicProto] = "tcp"
+	} else {
+		labels[aefLabelPublicProto] = "both"
+	}
+
+	// TODO(dev): collect max resource sizes: cpu, memory, disk.
+	//   Resources.Cpu
+	//   Resources.DiskGb
+	//   Resources.MemoryGb
+	//   Resources.Volumes[0].Name
+	//   Resources.Volumes[0].SizeGb
+	//   Resources.Volumes[0].VolumeType
+
+	// TODO: do we need to support multiple forwarded ports? How to choose?
+	// Extract target address in the form of the VM public IP and forwarded port.
+	re := regexp.MustCompile("([0-9]+)(/.*)")
+	port := re.ReplaceAllString(version.Network.ForwardedPorts[0], "$1")
+	targets := []string{fmt.Sprintf("%s:%s", instance.VmIp, port)}
+
+	// Construct a record for the Prometheus file service discovery format.
+	// https://prometheus.io/docs/operating/configuration/#<file_sd_config>
+	values := map[string]interface{}{
+		"labels":  labels,
+		"targets": targets,
+	}
+	return values
+}
+
+// listVms walks through every AppEngine service, looks at every serving
+// version. listVms returns a list of every running instance in a form suitable
+// for export to a prometheus service discovery file.
+func listVms(client *discovery) ([]map[string]interface{}, error) {
```

Comment:
Nesting is too deep.  Ok for quick start, but would prefer you extract functions in the next one or  two PRs.